jax/numpy/lax_numpy.py

```diff
@@ -1153,29 +1152,38 @@ def unravel_index(indices, shape):
 
 
 @_wraps(np.squeeze)
-def squeeze(a, axis=None):
+def squeeze(a, axis: Union[int, Tuple[int, ...]] = None):
   shape_a = shape(a)
   if axis is None:
     if 1 not in shape_a:
       return a
-    newshape = [d for d in shape_a if d != 1]
+    indices = tuple(0 if d == 1 else slice(None) for d in shape_a)
   else:
     if isinstance(axis, int):
       axis = (axis,)
-    axis = frozenset(_canonicalize_axis(i, ndim(a)) for i in axis)
-    if _any(shape_a[a] != 1 for a in axis):
+    axis_set = frozenset(_canonicalize_axis(i, ndim(a)) for i in axis)
+    if not axis_set:
+      return a
+    if _any(shape_a[a] != 1 for a in axis_set):
       raise ValueError("cannot select an axis to squeeze out which has size "
                        "not equal to one")
-    newshape = [d for i, d in enumerate(shape_a)
-                if d != 1 or i not in axis]
-  return lax.reshape(a, newshape)
+    indices = tuple(0 if i in axis_set else slice(None) for i in range(len(shape_a)))
+  # TODO: consider using lax.slice instead here?
```

Comment:
It seems to me like it is using too heavyweight a primitive (`gather`) to collapse away degenerate dimensions. It's pretty complicated to understand gather's semantics and I'm not even completely sure XLA will optimize away this kind of degenerate gather. (Hopefully it will.)

Perhaps we should just add a new lax-level `squeeze` primitive? You could implement this via `Reshape` in its translation rule, but it would preserve the necessary structure at the `lax` level.

Reply:
Good idea! I added `lax.squeeze`, with an implementation in terms of `reshape`. There's a nice duality between `squeeze` and `broadcast_in_dim` (for insertion of size 1 dimensions).